proof-producer/bin/proof-producer/include/nil/proof-generator/assignment_table_writer.hpp

```diff
@@ -0,0 +1,162 @@
+                    assert(integer_container.write(write_iter, char_array.size()) ==
+                        nil::marshalling::status_type::success);
+
+                    out_.write(reinterpret_cast<char*>(char_array.data()), char_array.size());
+                }
+
+                /**
+                * @brief Write zero value serialized via crypto3 marshalling into output
+                * stream.
+                */
+                inline void write_zero_field() {
+                    using empty_field = std::array<std::uint8_t, MarshallingField().length()>;
+                    
+                    empty_field field{};
+                    out_.write(reinterpret_cast<char*>(field.data()), field.size());
+                }
+
+                /**
+                * @brief Write field element into output stream.
+                */
+                void write_field(const FieldType& input) {
+                    MarshallingField field_container(input);
+                    std::array<std::uint8_t, field_container.length()> char_array{};
+                    auto write_iter = char_array.begin();
+                    assert(field_container.write(write_iter, char_array.size()) ==
+                        nil::marshalling::status_type::success);
+
+                    out_.write(reinterpret_cast<char*>(char_array.data()), char_array.size());
+                }
+
+
+                /**
+                * @brief Write table column to output stream padding with zeroes up to fixed number of values.
+                */
+                // template<typename Endianness, typename ArithmetizationType, typename ColumnType>
+                void write_vector_value(const std::size_t padded_rows_amount, const Column& table_col) {
+                    for (std::size_t i = 0; i < padded_rows_amount; i++) {
+                        if (i < table_col.size()) {
+                            write_field(table_col[i]);
+                        } else {
+                            write_zero_field();
+                        }
+                    }
+                }
+
+            public:
+                assignment_table_writer(std::ostream& out) : out_(out) {}
+                assignment_table_writer(assignment_table_writer&&) = delete;
+                assignment_table_writer(const assignment_table_writer&) = delete;
+                assignment_table_writer& operator=(const assignment_table_writer&) = delete;
+
+                void write_binary_assignment(const AssignmentTable& table, const AssignmentTableDescription& desc) {
+                    std::uint32_t public_input_size = table.public_inputs_amount();
+                    std::uint32_t witness_size = table.witnesses_amount();
+                    std::uint32_t constant_size = table.constants_amount();
+                    std::uint32_t selector_size = table.selectors_amount();
+                    std::uint32_t usable_rows_amount = desc.usable_rows_amount;
```

Comment:
better calculate `usable_rows_amount`
```
std::uint32_t max_witness_size = 0;
    for (std::uint32_t i = 0; i < witness_size; i++) {
        max_witness_size = std::max(max_witness_size, table.witness_column_size(i));
    }
    for (std::uint32_t i = 0; i < constant_size; i++) {
        max_constant_size = std::max(max_constant_size, table.constant_column_size(i));
    }
    for (std::uint32_t i = 0; i < selector_size; i++) {
        max_selector_size = std::max(max_selector_size, table.selector_column_size(i));
    }
    std::uint32_t usable_rows_amount =
        std::max({max_witness_size, max_public_inputs_size, max_constant_size, max_selector_size});
```
in case of generate assignment_table we'll have no description

Reply:
Actually this code caused initial trouble with incorrect writing. Seems like it for every table part it returns already padded size instead of the actual one (which stored in the description). Will take a look how to force it to return non-padded sizes

Reply:
sorry, you are right, here better use description